librubyfmt/src/render_queue_writer.rs

```diff
@@ -18,21 +18,59 @@ impl RenderQueueWriter {
 
     pub fn write<W: Write>(self, writer: &mut W) -> io::Result<()> {
         let mut accum = Intermediary::new();
+        let mut heredocs = vec![];
         #[cfg(debug_assertions)]
         {
             debug!("first tokens {:?}", self.tokens);
         }
-        Self::render_as(&mut accum, self.tokens);
+        Self::render_as(&mut accum, self.tokens, &mut heredocs);
         Self::write_final_tokens(writer, accum.into_tokens())
     }
 
-    fn render_as(accum: &mut Intermediary, tokens: Vec<ConcreteLineTokenAndTargets>) {
+    fn render_as(accum: &mut Intermediary, tokens: Vec<ConcreteLineTokenAndTargets>, heredocs: &mut Vec<HeredocString>) {
         for next_token in tokens.into_iter() {
             match next_token {
                 ConcreteLineTokenAndTargets::BreakableEntry(be) => {
-                    Self::format_breakable_entry(accum, be)
+                    Self::format_breakable_entry(accum, be, heredocs)
+                }
+                ConcreteLineTokenAndTargets::HeredocString(hs) => {
+                    heredocs.push(hs);
+                    #[cfg(debug_assertions)]
+                    {
+                        debug!("render_as: got heredoc {}", heredocs.len())
+                    }
+                }
+                ConcreteLineTokenAndTargets::ConcreteLineToken(ConcreteLineToken::LTStringContent {content}) => {
+                    let parts: Vec<&str> = content.splitn(2, '\n').collect();
+
+                    if !parts[0].is_empty() {
+                        accum.push(ConcreteLineToken::LTStringContent {
+                            content: parts[0].to_string()
+                        });
+                    }
+                    if parts.len() == 2 {
+                        accum.push(ConcreteLineToken::HardNewLine);
+                        Self::format_heredocs(accum, heredocs);
+
+                        if !parts[1].is_empty() {
+                            accum.push(ConcreteLineToken::LTStringContent {
+                                content: parts[1].to_string()
+                            });
+                        }
+                    }
+
+                }
```

Comment:
:thought_balloon: This match branch was a late attempt to solve some of the issues mentioned. In a previous iteration of this code, the splitting of string tokens into strings and newlines was handled in `emit_string_content` in ParserState, which I think was a more elegant way to do this.